iree/compiler/Codegen/LLVMGPU/Verifiers.cpp

```diff
@@ -0,0 +1,160 @@
+    unsigned bytesSize = type.getElementType().getIntOrFloatBitWidth() / 8;
+
+    // Input shape sizes: A [ M x K],  B [ K x N]
+    unsigned totalSharedMemSizeBytes =
+        (firstLevelTileSizes[0] * firstLevelTileSizes[2] +
+         firstLevelTileSizes[1] * firstLevelTileSizes[2]) *
+        bytesSize;
+
+    if (totalSharedMemSizeBytes > kSharedMemSizeBytes) {
+      return op->emitOpError("expected shared memory usage <= 64Kb for ")
+             << pipelineName << ", got " << totalSharedMemSizeBytes;
+    }
+  }
+  return success();
+}
+
+LogicalResult verifyGPUMatmulTensorCorePipeline(
+    Operation *op, IREE::Codegen::LoweringConfigAttr loweringConfig,
+    IREE::Codegen::TranslationInfoAttr translationInfo,
+    ArrayRef<int64_t> workgroupSize) {
+  auto pipeline =
+      IREE::Codegen::DispatchLoweringPassPipeline::LLVMGPUMatmulTensorCore;
+  StringRef pipelineName = stringifyEnum(pipeline);
+  if (workgroupSize.empty()) {
+    return op->emitOpError("expected workgroup size for GPU pipelines");
+  }
+
+  // Verify the total workgroup size is <= 1024
+  int64_t totalWorkgroupSize =
+      workgroupSize[0] * workgroupSize[1] * workgroupSize[2];
+  if (totalWorkgroupSize > 1024) {
+    return op->emitOpError("expected workgroup size to be <=1024 for ")
+           << pipelineName << ", got " << totalWorkgroupSize;
+  }
+
+  // Verify that the workgroup X dimension is 32 aligned
+  if (workgroupSize[0] % 32 != 0) {
+    return op->emitOpError("workgroup size is not 32 aligned for ")
+           << pipelineName << ", got " << workgroupSize[0];
+  }
+
+  // Verify the workgroup.z component should always be 1
+  if (workgroupSize[2] != 1) {
+    return op->emitOpError("expected workgroup z component to be 1 for ")
+           << pipelineName << ", got " << workgroupSize[2];
+  }
+
+  // Verify the second level of tiling = first level tile size divided by the
+  // warps per workgroup size
+  SmallVector<int64_t> firstLevelTileSizes =
+      loweringConfig.getTileSizeVals(kWorkgroupTileLevel);
+  std::array<int64_t, 3> warpsPerWorkgroup = {
+      workgroupSize[0] / kWarpSize, workgroupSize[1], workgroupSize[2]};
+  SmallVector<int64_t, 3> secondLevelTileSizes;
+  for (int i = 0; i < firstLevelTileSizes.size(); ++i) {
+    secondLevelTileSizes.push_back(firstLevelTileSizes[i] /
+                                   warpsPerWorkgroup[i]);
+  }
+
+  // Verify the TensorCore size divides the second level tile size
+  SmallVector<int64_t, 3> tensorCoreSize({16, 16, 8});
+  if (secondLevelTileSizes[0] % tensorCoreSize[0] != 0 ||
+      secondLevelTileSizes[1] % tensorCoreSize[1] != 0 ||
+      secondLevelTileSizes[2] % tensorCoreSize[2] != 0) {
+    return op->emitOpError(
+               "tensorcore size doesn't factor into second level tile size "
+               "for ")
+           << pipelineName;
+  }
+
+  // For linalg::MatmulOp, verify the first level tile size divides the matmul
```

Comment:
@ThomasRaoux In our notes, we didn't list a shared memory size verifier for the TensorCore pipeline. Should there be one here as well?

Reply:
Yes it should be the same. This is a hw/driver limit